### DISCUSSION TITLE
Potential race condition on async atoms?

### DISCUSSION THREAD

**Discussion by gimelfarb** [pmndrs/jotai#1761] [Q&A]
Before considering this a bug, it would be good to understand the existing reasons for this logic. I found this while debugging an issue where a derived atom didn't seem to refresh, even those the base atom was being changed.

The scenario is this:

```js
const baseAtom = atom(INIT);
const derivedAtom = atom(
  async (get) => {
    const value = get(baseAtom);
    if (value === INIT) return await externalFn();
    return value;
  },
  async (get, set, val) => set(baseAtom, val),
);

const writeAtom = atom(null, async (get, set) => {
  const value = await get(derivedAtom);
  if (condition(value)) set(derivedAtom, differentValue);
});
```

I don't think the arrangement of the code matters much. The important thing is performing a `set(derivedAtom, ...)` straight after `await get(derivedAtom)`.

Problematic code is here:
https://github.com/pmndrs/jotai/blob/f43ea723e4c0901ba428880b0ec0accd8f630e8d/src/vanilla/store.ts#L318-L341

Because jotai attaches a `.finally()` handler to the promise, it executes **AFTER** the promise value resolves in the `.then()` handler. That means that if we `await get(derivedAtom)`, then the code straight after that executes before the internal jotai `.finally()` block - in this case, we are doing `set(drivedAtom, ...)`. However, `.finally()` block then **overwrites** it with `setAtomValue(...)`, which means that `derivedAtom` will never re-calculate, even though `baseAtom` changed.

Now, I don't know of the reasons why the `.finally()` block is arranged in this manner in jotai codebase. While this may not be a common usage scenario, it still seems to be semantically incorrect. It's a **race condition** - `setAtomValue(...)` is called AFTER atom's promise value resolves, which can lead to unexpected behaviour such as this.

If I were to propose a fix, I'd move `setAtomValue(...)` out of `.finally()` block, and just before `resolve` or `reject` in the `.then()` block.

Am I right to see this as a potential issue?

As of now, in my code, I've done a workaround similar to this:

```js
const writeAtom = atom(null, async (get, set) => {
  const value = await get(derivedAtom);
  const setAsync = (a, ...args) => new Promise((resolve) => setTimeout(() => resolve(set(a, ...args)), 0));
  if (condition(value)) setAsync(derivedAtom, differentValue);
});
```

The workaround just ensures that I don't call `set` straight away after the `await get(derivedAtom)`, which allows the internal `.finally()` block to execute first.

**Comment by dai-shi** [comment]
Nice catch! And thanks for the investigation.
Yeah, that seems bad. (btw, we we didn't support this use case in v1. Did you use `unstable_promise` before?)

Yeah, removing the `.finally()` block makes sense. Can you give it a try? As far as I remember, it made some tests to fail. You may find other bugs then. Thanks for your help.

**Reply by gimelfarb** [reply]
@dai-shi Thank you for taking a look! I haven't actually used Jotai v1. I decided to give it a try when v2 was already available via `jotai/vanilla` import, and so I jumped on the new bandwagon straight away. So my Jotai experience is pretty recent :)

Yes, I can give it a try, fork the repo and see if the change fails any existing tests. Will also give me a chance to come up with a minimal reproducible scenario.

**Reply by gimelfarb** [reply]
@dai-shi Created PR for the fix: https://github.com/pmndrs/jotai/pull/1768

Includes a unit test that fails with existing code, and succeeds with the fix. Other tests seem to pass as well :)
